src/internal/task/etcd_task_service.go

```diff
@@ -0,0 +1,340 @@
+package task
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	etcd "github.com/coreos/etcd/clientv3"
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	col "github.com/pachyderm/pachyderm/v2/src/internal/collection"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errutil"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachhash"
+	"github.com/pachyderm/pachyderm/v2/src/internal/uuid"
+	"github.com/pachyderm/pachyderm/v2/src/internal/watch"
+)
+
+const (
+	groupPrefix = "/group"
+	taskPrefix  = "/task"
+	claimPrefix = "/claim"
+)
+
+type Service interface {
+	// TODO: Should group ID parameter?
+	Doer(ctx context.Context, namespace, groupID string, cb func(context.Context, Doer) error) error
```

Comment:
Why does this need a context?, why does it need to return an error, and why does it take a callback instead of returning the doer.

Reply:
The callback in particular is going to make initialization very difficult.

A system (like PFS) should be able to receive a Service, and then spawn Doers or makers and pass them to the master and compaction worker.  As written, the entire lifetime of the compaction worker or pfs master would have to be limited to inside the callback.

Reply:
We don't want the doers to be scoped like that because we want the groups to be at a lower level of granularity (i.e. the pfs master would create a doer for each repo or commit (it is commit now) being processed rather than a doer for all of the commits that will be processed). The PFS / PPS code has been migrated over if you want to see what it looks like. We need to return a callback and close the renewer, which is why I went with this pattern, but we could also handle that by exposing a close method and returning a context. The callback pattern just seems safer / cleaner for this.